server/handlers/cartLock.js

```diff
@@ -0,0 +1,18 @@
+const cartLocks = new Map();
+
+exports.acquireCartLock = async function (userId) {
+    return new Promise((resolve, reject) => {
+        const lock = setInterval(() => {
+            if (!cartLocks.has(userId)) {
+                cartLocks.set(userId, true);
+                clearInterval(lock);
```

Comment:
This is used for avoiding thread conflict.
Every time the request is made, it will acquire a lock for this user's cart. If `cartLocks` isn't held by another user's id, it will lock the current cart and process the request until finishes. When this request is finished, it will release the lock. If `cartLocks` is held by another user, it will keep asking for the lock every 10ms until the lock is released by the previous user.
When the new request is under-processed, it will stop the continuous asking by using `clearInterval(lock) `.

Reply:
> why use clearInterval here?

This is used for avoiding thread conflict.
Every time the request is made, it will acquire a lock for this user's cart. If `cartLocks` isn't held by another user's id, it will lock the current cart and process the request until finishes. When this request is finished, it will release the lock.
If `cartLocks` is held by another user, it will keep asking for the lock every 10ms until the lock is released by the previous user.
When the new request is under-processed, it will stop the continuous asking by using `clearInterval(lock)`.